### PR TITLE
Revising explainability of linear models

### PR DESCRIPTION
**Pull request by astonzhang**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by astonzhang** [comment]
1. For notation consistency with https://d2l.ai/chapter_linear-networks/linear-regression.html, N -> n, p -> d
2. The bar plot is not in the svg format (not as sharp as the plot in section 1).
3. "which can be measured by the Pearson, Kendall, or Spearman correlation coefficient". A textbook can be more self-contained. Define them in math and explain special values (e.g., -1, 0, 1).
4. "if two tokens contribute equally to all possible coalitions, their contribution values are the same" What do token mean? Features? Can we simplify "coalitions" (undefined) to make it easier to understand?
5. Can we use code to show more examples, e,g, (i) issues of correlated features in explainability; (ii) how we can fix issues (can be any issue, such as correlated features, unitless features) to show *improved* (from bad to good) explainability results. For example, we only have good results in feature importance with unitless features - can we show bad results before unitless features so readers can see *improvement*? This will make our story more engaging.

**Comment by cheungdaven** [comment]
> @cheungdaven
> 
>     1. For notation consistency with https://d2l.ai/chapter_linear-networks/linear-regression.html, N -> n, p -> d
> 
>     2. The bar plot is not in the svg format (not as sharp as the plot in section 1).
> 
>     3. "which can be measured by the Pearson, Kendall, or Spearman correlation coefficient". A textbook can be more self-contained. Define them in math and explain special values (e.g., -1, 0, 1).
> 
>     4. "if two tokens contribute equally to all possible coalitions, their contribution values are the same" What do token mean? Features? Can we simplify "coalitions" (undefined) to make it easier to understand?
> 
>     5. Can we use code to show more examples, e,g, (i) issues of correlated features in explainability; (ii) how we can fix issues (can be any issue, such as correlated features, unitless features) to show _improved_ (from bad to good) explainability results. For example, we only have good results in feature importance with unitless features - can we show bad results before unitless features so readers can see _improvement_? This will make our story more engaging.

Replies to questions:
1. will do that. 
2. I basically used the same plot function defined in section 1, it is weird.
3. Do you mean define Pearson mathematically?
4. will do that.
5. (1) Our data does not high correlated features, one way is to generate some fake data, but the model might get very bad results on which it can be meaningless to make any explanations.  (2) Explaining with the raw features is not BAD but faulty.

File: chapter_explainability/explainability-linear.md [hunk 1]
```diff
@@ -72,64 +81,56 @@ d2l.plot_hbar(data.feat_col, regr.coef_ * data.X_train.std(axis=0))
...
+If two features depend on or associate with each another, we call them *correlated features* (also known as *multicollinearity* or *collinearity*). Correlated features do not bring additional information (or just very little). Instead, they might increase the model complexity and make it difficult to find a unique solution for linear models. Also, correlated features make the coefficient interpretation unreliable because they tend to change in unison. When features are correlated, changes in one are associated with variations in the other. The stronger the correlation, the more difficult it is to change one feature without changing the other. Therefore, for any two highly correlated features, practitioners may keep only one of them in the dataset.
```
Comment (astonzhang) [inline chapter_explainability/explainability-linear.md:84]: remove multicolinear

File: chapter_explainability/explainability-linear.md [hunk 2]
```diff
@@ -72,64 +81,56 @@ d2l.plot_hbar(data.feat_col, regr.coef_ * data.X_train.std(axis=0))
...
-\frac{\text{odds} (x_j^{(i)}+1)}{\text{odds}} &= \frac{\exp (b + w_1 x_1^{(i)} + ...+w_j (x_j^{(i)}+1)+...+ w_p x_p^{(i)})}{\exp (b + w_1 x_1^{(i)} + ...+w_j x_j^{(i)}+...+ w_p x_p^{(i)})} \\
-&= \exp (w_j (x_j^{(i)}+1) - w_j x_j^{(i)}) = \exp(w_j)
-\end{aligned}
+We will use *odds ratio* to explain logistic regression models, which is the ratio between two events: increasing $x_j$ by one unit vs. keeping it fixed:
+
+$$\frac{\text{odds} (x_j+1)}{\text{odds}} = \frac{\exp (b + w_1 x_1 + \ldots+w_j (x_j+1)+\ldots+ w_p x_p)}{\exp (b + w_1 x_1 + \ldots+w_j x_j+\ldots+ w_p x_p)}  = \exp(w_j).
 $$
 
-Let's compute the odds of the logistic regression model.
+Odds ratio interprets logistic regression handily: increasing $x_j$ by one unit will scale the odds ratio by a factor of $\exp(w_j)$. Let's compute them for our logistic regression model in heart disease prediction.
```
Comment (astonzhang) [inline chapter_explainability/explainability-linear.md:118]: the odds ratio -> the odds of XXX (to be concrete)

File: chapter_explainability/explainability-linear.md [hunk 3]
```diff
@@ -72,64 +81,56 @@ d2l.plot_hbar(data.feat_col, regr.coef_ * data.X_train.std(axis=0))
...
+Now we can explain our logistic regression model using odds ratio for each feature.
+For example, on one hand, an increase in a patient's age nearly does not change the odds of having heart disease vs. not having heart disease (the odds ratio is around $1$); on the other hand, for chest pain type (0: typical angina, 1: atypical angina, 2: non-anginal pain, 3: asymptomatic), changing the categorical type from "typical angina" to "atypical angina" will at least double the heart disease odds ratio (the odds ratio is over $2$). Note that such interpretations are on the model behavior and do not imply real-world causality.
```
Comment (astonzhang) [inline chapter_explainability/explainability-linear.md:126]: at least double the heart disease odds ratio
-> not odds ratio

**Pull request merged**